### PR TITLE
Store StagingWorkflow in the backend

### PR DESCRIPTION
**Pull request by DavidKang**
StagingWorkflow is stored in the backend under `_project` directory with
the name `_staging_workflow`. In this special file, we store a minimal
info of the staging_workflow.

### CODE REVIEW

File: src/api/app/models/staging_workflow.rb [hunk 1]
```diff
@@ -32,6 +32,33 @@ def ignored_requests
...
+                                   state: staged_request.state,
+                                   creator: staged_request.creator }
+
+            builder_staging_project.request(request_attributes)
+          end
+        end
+      end
+    end
+
+    builder.to_xml(save_with: Nokogiri::XML::Node::SaveOptions::NO_DECLARATION | Nokogiri::XML::Node::SaveOptions::FORMAT)
```
Comment (hennevogel) [inline src/api/app/models/staging_workflow.rb]: Why don't you do all of this in a view?

File: src/api/app/controllers/webui/staging_workflows/staging_projects_controller.rb [hunk 2]
```diff
@@ -14,6 +14,7 @@ def create
         staging_project = @staging_workflow.staging_projects.build(name: staging_project_name)
 
         if staging_project.valid? && staging_project.store
+          @staging_workflow.write_to_backend
```
Comment (hennevogel) [inline src/api/app/controllers/webui/staging_workflows/staging_projects_controller.rb]: Why don't you do this in a callback?

File: src/api/app/models/staging_workflow.rb [hunk 3]
```diff
@@ -32,6 +32,33 @@ def ignored_requests
     BsRequest.none # TODO: define this method
   end
 
+  def to_xml
+    builder = Nokogiri::XML::Builder.new
+    staging_workflow_attributes = { project: project.name,
+                                    who: User.current_login,
```
Comment (coolo) [inline src/api/app/models/staging_workflow.rb]: this who is strange - who saved the workflow last is part of the _history of _project, no need to repeat it within the xml. if nothing changed, the xml should stay the same - who didn't change it, is of no interet :)

Comment (coolo) [reply src/api/app/models/staging_workflow.rb]: same for when actually - also in _history

File: src/api/app/models/staging_workflow.rb [hunk 4]
```diff
@@ -32,6 +32,33 @@ def ignored_requests
...
+                                    when: updated_at }
+
+    builder.staging_workflow(staging_workflow_attributes) do |builder_staging_workflow|
+      staging_projects.each do |staging_project|
+        builder_staging_workflow.staging_project(name: staging_project.name) do |builder_staging_project|
+          staging_project.staged_requests.each do |staged_request|
+            request_attributes = { number: staged_request.number,
```
Comment (coolo) [inline src/api/app/models/staging_workflow.rb]: I don't think the staged requests belong in the backend. this would be *super* noisy on _history

Comment (hennevogel) [reply src/api/app/models/staging_workflow.rb]: But that's the only thing that changes...

Comment (coolo) [reply src/api/app/models/staging_workflow.rb]: well, in daily routine yes - but it will be basically impossible to find any other changes in Factory's _history.

So storing the selection history within the staging project might make sense - but I would limit _staging_workflow to changes to the *workflow*, not the work done within

File: src/api/app/models/staging_workflow.rb (hunk 3 above)
Comment (DavidKang) [reply src/api/app/models/staging_workflow.rb]: I will remove it

File: src/api/app/controllers/webui/staging_workflows/staging_projects_controller.rb (hunk 2 above)
Comment (DavidKang) [reply src/api/app/controllers/webui/staging_workflows/staging_projects_controller.rb]: let's see :smile_cat:

File: src/api/app/models/staging_workflow.rb (hunk 4 above)
Comment (hennevogel) [reply src/api/app/models/staging_workflow.rb]: Okay, sounds reasonable to me. @DavidKang please remote requests.

File: src/api/app/controllers/webui/staging_workflows/staging_projects_controller.rb (hunk 2 above)
Comment (coolo) [reply src/api/app/controllers/webui/staging_workflows/staging_projects_controller.rb]: I would really encourage putting it in the model - we had the problem with attributes being stored only from the API controller and it created quickly a mess, as the webui controllers were refactored (not to use API anymore).

Doing it in the model created different problems - especially with cassettes, but I leave you that sad story for a rainy night

File: src/api/app/models/project.rb [hunk 5]
```diff
@@ -1750,6 +1753,15 @@ def has_local_distribution(project_name, repository)
         linked_repository.name == repository
     end
   end
+
+  def update_staging_workflow
+    transaction do
```
Comment (ChrisBr) [inline src/api/app/models/project.rb]: This is still wrapped in the ``save`` transaction.
https://apidock.com/rails/ActiveRecord/Callbacks/after_save

Comment (DavidKang) [reply src/api/app/models/project.rb]: @ChrisBr thanks :smile_cat:

File: src/api/app/models/staging_workflow.rb (hunk 1 above)
Comment (hennevogel) [reply src/api/app/models/staging_workflow.rb]: I mean doing the layout of the xml in a `app/views/staging_workflow.xml.builder` instead of using builder inside the model. then we can also make use of this for showing etc.

Comment (DavidKang) [reply src/api/app/models/staging_workflow.rb]: uhhmmm, I will try it :smile_cat:. Thanks

Comment (hennevogel) [reply src/api/app/models/staging_workflow.rb]: It the better pattern for the xml rendering IMHO. see for instance the status/required checks...

Comment (hennevogel) [reply src/api/app/models/staging_workflow.rb]: > but rendering a view in a model 

Hey would you so that? You would also not build HTML inside a model, would you?

Comment (hennevogel) [reply src/api/app/models/staging_workflow.rb]: > but rendering a view in a model 

Hey would you so that? You would also not build HTML inside a model, would you?

File: src/api/app/models/project.rb [hunk 6]
```diff
@@ -28,6 +28,9 @@ class Project < ApplicationRecord
     @commit_opts = {}
   end
 
+  after_save :update_staging_workflow
+  after_destroy :update_staging_workflow
```
Comment (coolo) [inline src/api/app/models/project.rb]: what's the point of storing anything after destroy? If the project is gone, the staging workflow is too

Comment (DavidKang) [reply src/api/app/models/project.rb]: @coolo, this `after_destroy` only affects `staging_projects`.

Comment (coolo) [reply src/api/app/models/project.rb]: I see

File: src/api/app/models/staging_workflow.rb (hunk 1 above)
Comment (DavidKang) [reply src/api/app/models/staging_workflow.rb]: Okey, I changed the code to use `CanRenderModel` mixing instead of `Nokogiri`

**Comment by coolo** [comment]
writing to that file outside of the model should be plainly forbidden IMO.

**Comment by coolo** [comment]
how about?

```
diff --git a/src/api/config/routes.rb b/src/api/config/routes.rb
index bb4204d7f..a49cc73c0 100644
--- a/src/api/config/routes.rb
+++ b/src/api/config/routes.rb
@@ -768,6 +768,7 @@ OBSApi::Application.routes.draw do
 
     get 'source/:project/:package/:filename' => :get_file, constraints: cons, defaults: { format: 'xml' }
     delete 'source/:project/:package/:filename' => :delete_file, constraints: cons
+    put 'source/:project/_project/_staging_workflow' => :just_dont, constraints: cons
     put 'source/:project/:package/:filename' => :update_file, constraints: cons

**Comment by adrianschroeter** [comment]
That would be just one out of many ways to work on it. And why should it be forbidden, if you add it here it is obvious that it is also the API to create/modify it, no?

**Comment by hennevogel** [comment]
[This is](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/controllers/source_controller.rb#L335) how we forbid other underscore files. An API for creating/changing this is not what we are going to do now.

File: src/api/app/models/project.rb [hunk 7]
```diff
@@ -1750,6 +1753,13 @@ def has_local_distribution(project_name, repository)
         linked_repository.name == repository
     end
   end
+
+  def update_staging_workflow
```
Comment (hennevogel) [inline src/api/app/models/project.rb]: This is not really updating it right? Confusing name.

File: src/api/spec/models/staging_workflow_spec.rb [hunk 8]
```diff
@@ -1,7 +1,7 @@
 require 'rails_helper'
 
-RSpec.describe StagingWorkflow, type: :model do
-  let(:project) { create(:project_with_package) }
+RSpec.describe StagingWorkflow, type: :model, vcr: true do
```
Comment (hennevogel) [inline src/api/spec/models/staging_workflow_spec.rb]: Why is this necessary?

Comment (DavidKang) [reply src/api/spec/models/staging_workflow_spec.rb]: If you are talking about the VCR, is because we store the XML file in the backend when we create/update a staging_workflow

File: src/api/app/models/project.rb (hunk 7 above)
Comment (DavidKang) [reply src/api/app/models/project.rb]: Maybe `update_staging_workflow_on_backend`?

**Comment by adrianschroeter** [comment]
@hennevogel this is on package name level and not on file level.

The api it needs also be writeable anyway in the end, it is bad idea to add cruel hacks to workaround some situations (and for sure not all). It shows that the design should have been documented first before hacking ....

**Comment by coolo** [comment]
does `if @file == '_attribute'` ring any bells, @adrianschroeter ?

**Comment by coolo** [comment]
using a file to create staging projects is a bad API anyway. So let's talk about an API to modify staging workflows independent of _project

File: src/api/app/models/project.rb (hunk 7 above)
Comment (hennevogel) [reply src/api/app/models/project.rb]: sure why not :smile:

**Comment by adrianschroeter** [comment]
_attribute is a very special case, because it can not be bind to the package model container of "_project" (since it is project model). reading works different there anyway. But this not true for a regular source file.

dunno if a file in _project makes sense for this, but it must be configurable via api anyway and there should be two ways for same content. So, yes, better define first how to create/modify/delete it via api first before defining the place.

**Comment by hennevogel** [comment]
> So let's talk about an API to modify staging workflows independent of _project

We will do that for sure, I don't like having UI or API only features either. But not now, in a later iteration.

**Comment by adrianschroeter** [comment]
@hennevogel fine if you don't want to do it now, but then don't store this file either now. It only cries for problems, stuff will be out of sync, people may use it already because they can see it.... 

I was only against to it partly and creating all these problems.

**Comment by hennevogel** [comment]
>  but then don't store this file either now

We can also just "forbid" writing to this file like we do for _attribute for now.

File: src/api/spec/models/staging_workflow_spec.rb (hunk 8 above)
Comment (DavidKang) [reply src/api/spec/models/staging_workflow_spec.rb]: I will check it

File: src/api/app/models/project.rb [hunk 9]
```diff
@@ -1750,6 +1753,13 @@ def has_local_distribution(project_name, repository)
         linked_repository.name == repository
     end
   end
+
+  def update_staging_workflow_on_backend
+    return if staging_workflow_id.nil?
```
Comment (Ana06) [inline src/api/app/models/project.rb]: :see_no_evil: `return unless staging_workflow_id`

Comment (dmarcoux) [reply src/api/app/models/project.rb]: It's not exactly the same, but I guess in this case, it will be fine.

**Comment by DavidKang** [comment]
@hennevogel and @bgeuken could you review it again?

**Comment by hennevogel** [comment]
code* and hakiri hang. Coverage is fine, see coveralls and there are no security warnings on hakiri for this commit. Merging...
